aten/src/ATen/native/UpSampleNearest3d.cpp

```diff
@@ -0,0 +1,294 @@
+          const scalar_t* pos2 =
+              &odata
+                  [d2 * output_height * output_width + h2 * output_width + w2];
+
+          for (int64_t c = 0; c < channels; ++c) {
+            pos1[0] += pos2[0];
+            pos1 += input_depth * input_height * input_width;
+            pos2 += output_depth * output_height * output_width;
+          }
+        }
+      }
+    }
+    return;
+  }
+
+  for (int64_t d2 = 0; d2 < output_depth; ++d2) {
+    const int64_t d1 =
+        nearest_neighbor_compute_source_index(depth_scale, d2, input_depth);
+
+    for (int64_t h2 = 0; h2 < output_height; ++h2) {
+      const int64_t h1 =
+          nearest_neighbor_compute_source_index(height_scale, h2, input_height);
+
+      for (int64_t w2 = 0; w2 < output_width; ++w2) {
+        const int64_t w1 =
+            nearest_neighbor_compute_source_index(width_scale, w2, input_width);
+        scalar_t* pos1 =
+            &idata[d1 * input_height * input_width + h1 * input_width + w1];
+        const scalar_t* pos2 =
+            &odata[d2 * output_height * output_width + h2 * output_width + w2];
+
+        for (int64_t c = 0; c < channels; ++c) {
+          pos1[0] += pos2[0];
+          pos1 += input_depth * input_height * input_width;
+          pos2 += output_depth * output_height * output_width;
+        }
+      }
+    }
+  }
+}
+
+static void upsample_nearest3d_out_cpu_template(
+    Tensor& output,
+    const Tensor& input_,
+    IntArrayRef output_size) {
+  int64_t output_depth = output_size[0];
+  int64_t output_height = output_size[1];
+  int64_t output_width = output_size[2];
+
+  int64_t nbatch = input_.size(0);
+  int64_t channels = input_.size(1);
+  int64_t input_depth = input_.size(2);
+  int64_t input_height = input_.size(3);
+  int64_t input_width = input_.size(4);
+
+  upsample_3d_shape_check(
+      input_,
+      Tensor(),
+      nbatch,
+      channels,
+      input_depth,
+      input_height,
+      input_width,
+      output_depth,
+      output_height,
+      output_width);
+
+  auto input = input_.contiguous();
+
+  output.resize_({nbatch, channels, output_depth, output_height, output_width});
+  output.zero_();
+
+  AT_ASSERT(
+      input_depth > 0 && input_height > 0 && input_width > 0 &&
+      output_depth > 0 && output_height > 0 && output_width > 0);
+
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.scalar_type(), "upsample_nearest3d", [&] {
+    auto* idata = input.data<scalar_t>();
+    auto* odata = output.data<scalar_t>();
+
+    upsample_nearest3d_out_frame<scalar_t>(
+        odata,
+        idata,
+        input_depth,
+        input_height,
+        input_width,
+        output_depth,
+        output_height,
+        output_width,
+        nbatch,
+        channels);
+  });
+}
+
+static void upsample_nearest3d_backward_out_cpu_template(
+    Tensor& grad_input,
+    const Tensor& grad_output_,
+    IntArrayRef output_size,
+    IntArrayRef input_size) {
```

Comment:
I think we now need checks that `output_size` and `input_size` are the correct length. These checks were previously autogenerated but since we accept the `IntArrayRef` directly you have to add them back.

Reply:
This applies to all of the functions. I won't comment on each of them.

Reply:
sounds good. just a question, I was search in a bunch of files that use `IntArrayRef` as a function parameter and I didn't find any check there. I am wondering if the signature in `native_functions.yaml` that specify the size of the array should already check that. eg. `- func: upsample_nearest3d_backward(Tensor grad_output, int[3] output_size, int[5] input_size, *, Tensor(a!) grad_input) -> Tensor(a!)`

Reply:
Ah, you might be right. It's easy enough to test; just try passing in a size that is too big/small and see what happens.

Reply:
thanks for the feedback @ezyang 
in my tests here using python, it seems there is no check for `size` variables. I will add this check on cpp files.